trie/committer.go

```diff
@@ -0,0 +1,284 @@
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package trie
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+	"golang.org/x/crypto/sha3"
+)
+
+// LeafChanSize is the size of the leafCh. It's a pretty arbitrary number, to allow
+// some paralellism but not incur too much memory overhead.
+const LeafChanSize = 200
+
+// Leaf represents a trie leaf value
+type Leaf struct {
+	size   int         // size of the rlp data (estimate)
+	hash   common.Hash // hash of rlp data
+	node   node        // the node to commit
+	vnodes bool        // set to true if the node (possibly) contains a valueNode
+}
+
+// committer is a type used for the trie Commit operation. A committer has some
+// internal preallocated temp space, and also a callback that is invoked when
+// leaves are committed. The leafs are passed through the `leafCh`,  to allow
+// some level of paralellism.
+// By 'some level' of parallelism, it's still the case that all leaves will be
+// processed sequentially - onleaf will never be called in parallel or out of order.
+type committer struct {
+	tmp sliceBuffer
+	sha keccakState
+
+	onleaf LeafCallback
+	leafCh chan *Leaf
+}
+
+// committers live in a global db.
+var committerPool = sync.Pool{
+	New: func() interface{} {
+		return &committer{
+			tmp: make(sliceBuffer, 0, 550), // cap is as large as a full fullNode.
+			sha: sha3.NewLegacyKeccak256().(keccakState),
+		}
+	},
+}
+
+// newCommitter creates a new committer or picks one from the pool, and
+// initializes the leafCh, if needed.
+// In case no onleaf-callback is provided, the committer does not
+// use a channel-based commit, but inlined.
+// Typically, the account trie is committed with a channel-based leaf-commit,
+// whereas storage tries are committed 'inline'.
+func newCommitter(onleaf LeafCallback) *committer {
+	h := committerPool.Get().(*committer)
+	h.onleaf = onleaf
+	if onleaf != nil {
+		h.leafCh = make(chan *Leaf, LeafChanSize)
+	}
```

Comment:
I think it would be cleaner to move the onleaf assignment into the if too (it gets reset in returnComitter, so no need to clean it up explicitly here):

```go
if onleaf != nil {
	h.onleaf = onleaf
	h.leafCh = make(chan *Leaf, LeafChanSize)
}
```